### DISCUSSION TITLE
Validating a nested object with a property from the parent

### DISCUSSION THREAD

**Discussion by markjaniczak** [typestack/class-validator#2074] [Questions & Answers]
There doesn't seem to be a straight forward way of validating a nested object based on a property from its parent or further up the tree. I'm looking for guidance on the best way to achieve this. 

For example, I have the following object from a form:

```json
{
    "name": "My Video Stream",
    "hasRedundancy": true,
    "source": {
        "type": "hls",
        "primary": "https://some-domain.com/index.m3u8",
        "secondary": "https://some-domain.com/backup.m3u8"
    }
}
```

The `secondary` property in the `source` object is only required if `has_redundancy` is `true`. So I write my classes like so:

```typescript
const SOURCE_TYPES = ['HLS', 'MP4'] as const

type SourceType = (typeof SOURCE_TYPES)[number]

class Source {
    type: SourceType
    primary: any
    secondary: any
}

class HLSSource extends Source {
    type: 'HLS'

    @IsUrl()
    primary: string

    @ValidateIf((obj) => obj.hasRedundancy)
    @IsUrl()
    secondary: string
}

class MP4Source extends Source {
  type: 'MP4'
  
  @IsUrl()
  primary: string

  @ValidateIf((obj) => obj.hasRedundancy)
  @IsUrl()
  secondary: any;
}

class CreateStreamRequest {
    @IsString()
    name: string

    @IsBoolean()
    hasRedundancy: boolean

    @ValidateNested({
      each: true
    })
    @Type(() => Source, {
      discriminator: {
        property: 'type',
        subTypes: [
          { name: 'MP4', value: MP4Source },
          { name: 'HLS', value: HLSSource}
        ]
      },
      keepDiscriminatorProperty: true
    })
    source: HLSSource | MP4Source
}
```

But `hasRedundancy` is not included in the plain object that the my `Source` is created from, only in the top level plain object that `CreateStreamRequest` is made from. I thought about using the groups feature to tackle this but that seems like double handling because I already have the data to pass to the condition in the original object. I see that the first parameter in the `Type` decorator provides some parameters but they are not documented anywhere.

**Comment by Iuriiiiiii** [comment]
Hi, i have this problem too... the parent information is a relevant data that can't be ignored within a complex custom contraint. What should we do?

**Reply by elovin** [reply]
I would recursively inject an object reference to the parent into all objects into a property with the `@Allow` & `@IsOptional` decorators.

E.g. using `@Transform()` on a custom property on the root Dto or some custom recursive function.

In case you are using NestJS you could do this with an interceptor and modify the request body directly, this interceptor could then also recursively remove the reference in case it ends up in the response for some reason.

**Comment by rinaldihtb** [comment]
Hi I'm also having this issue, just want to check if there's anyone that already have the solution ?

**Comment by usualdesigner** [comment]
The problem is @Type()'s discriminator resolution happens before your nested class ever sees the parent object, so hasRedundancy genuinely isn't available there. I'd write a custom class-level @ValidatorConstraint on CreateStreamRequest itself that checks object.hasRedundancy && !object.source.secondary and fails there - validating the cross-field rule at the level where both properties actually exist, rather than reaching up from the nested class.
